### PR TITLE
Homework3

### PR DESCRIPTION
**Pull request by oksanazah**
- connected to MongoDB Database
- created endpoints for movies and genres
- added data validation and error handling
- updated swagger documentation

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 import errorHandler from './middleware/error-handler';
 import healthChekcRouter from './routes/health-check.routes';
 import productsRouter from './routes/products.routes';
+import moviesRouter from './routes/movies.routes';
+import genresRouter from './routes/genres.routes';
 
 const app: Express = express();
 const PORT = 3000;
+const MONGODB_URI =
+  'mongodb+srv://oksanazaharchenko:4bIg0tLnbnA5edNR@cluster0.voxnl81.mongodb.net/NODE_JS_DB?retryWrites=true&w=majority';
```
Comment (vasylkivskyy) [inline src/index.ts]: Please do not push sensitive data to GitHub, use .env file for this, for example

File: src/routes/genres.routes.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ *         content:
+ *           application/json:
+ *             example:
+ *               name: New genre
+ *       400:
+ *         description: Bad Request
+ *       500:
+ *         description: Internal Server Error
+ */
+router.post('/', (req: Request, res: Response, next: NextFunction) => {
```
Comment (vasylkivskyy) [inline src/routes/genres.routes.ts]: Is it better to add validation middleware for endpoints (not 'GET') to validate each field that come in your paylaod, using for example Joi package or writting it by your own.

File: src/routes/genres.routes.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ *         description: Bad Request
+ *       404:
+ *         description: Genre not found
+ *       500:
+ *         description: Internal Server Error
+ */
+router.put('/:id', (req: Request, res: Response, next: NextFunction) => {
+  const genreId = req.params.id;
+
+  GenreModel.findByIdAndUpdate(genreId, req.body)
```
Comment (vasylkivskyy) [inline src/routes/genres.routes.ts]: It is good practice to use try catch blocks than promises with callbacks

**Review by vasylkivskyy — CHANGES_REQUESTED** [review]

File: src/models/models.ts [hunk 4]
```diff
@@ -1,4 +1,6 @@
-export interface Error {
```
Comment (vasylkivskyy) [inline src/models/models.ts:1]: better to move and rename this file to interfaces

File: src/index.ts (hunk 1 above)
Comment (oksanazah) [reply src/index.ts]: just wanted to make review easier, sorry, removed sensitive data

File: src/routes/genres.routes.ts (hunk 2 above)
Comment (oksanazah) [reply src/routes/genres.routes.ts]: used joi for validation

File: src/routes/genres.routes.ts (hunk 3 above)
Comment (oksanazah) [reply src/routes/genres.routes.ts]: rewrote, using try/catch

File: src/models/models.ts (hunk 4 above)
Comment (oksanazah) [reply src/models/models.ts:1]: remaned and moved this to interfaces

File: src/routes/genres.routes.ts [hunk 5]
```diff
@@ -143,14 +159,20 @@ router.put('/:id', (req: Request, res: Response, next: NextFunction) => {
...
-router.delete('/:id', (req: Request, res: Response, next: NextFunction) => {
-  const genreId = req.params.id;
+router.delete(
+  '/:id',
+  async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const genreId = req.params.id;
+
+      const deletedGenre: Genre | null =
+        await GenreModel.findByIdAndDelete(genreId);
```
Comment (vasylkivskyy) [inline src/routes/genres.routes.ts:169]: If deletedGenre will be null?
You need to add this checking and return appropriate status

File: src/routes/movies.routes.ts [hunk 6]
```diff
@@ -200,15 +218,21 @@ router.put('/:id', (req: Request, res: Response, next: NextFunction) => {
...
-router.delete('/:id', (req: Request, res: Response, next: NextFunction) => {
-  const movieId = req.params.id;
+router.delete(
+  '/:id',
+  async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const movieId = req.params.id;
+
+      const deletedMovie: Movie | null =
+        await MovieModel.findByIdAndDelete(movieId);
```
Comment (vasylkivskyy) [inline src/routes/movies.routes.ts:228]: The same situation as in deleting genres

File: src/routes/genres.routes.ts (hunk 5 above)
Comment (oksanazah) [reply src/routes/genres.routes.ts:169]: added check for null

File: src/routes/movies.routes.ts (hunk 6 above)
Comment (oksanazah) [reply src/routes/movies.routes.ts:228]: added check for null

File: src/routes/genres.routes.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+ *             example:
+ *               name: New genre
+ *       400:
+ *         description: Bad Request
+ *       500:
+ *         description: Internal Server Error
+ */
+router.post(
+  '/',
+  validateGenre,
```
Comment (vasylkivskyy) [inline src/routes/genres.routes.ts]: For improvements, you can  invoke validateGenre(genreSchema or movieSchema) - to not dublicate the same validate functions twice in middlewares folder

**Review by vasylkivskyy — APPROVED** [review]

**Pull request merged**